kv/kvserver: TestReplicateQueueTracingOnError failed

kv/kvserver.TestReplicateQueueTracingOnError [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsIbmCloudLinuxS390x_UnitTestsS390x/20861817?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsIbmCloudLinuxS390x_UnitTestsS390x/20861817?buildTab=artifacts#/) on release-25.3 @ [bd3f5d0dcbbcbd723c6d94a7bf9c7480cd184bf5](https://github.com/cockroachdb/cockroach/commits/bd3f5d0dcbbcbd723c6d94a7bf9c7480cd184bf5):


```
=== RUN   TestReplicateQueueTracingOnError
    test_log_scope.go:165: test logs captured to: /artifacts/tmp/_tmp/bb83881c24a4b108656da9268f863369/logTestReplicateQueueTracingOnError239442307
    replicate_queue_test.go:963: 
        	Error Trace:	pkg/kv/kvserver/replicate_queue_test.go:963
        	Error:      	Expect "error processing replica: 1 matching stores are currently throttled: [‹boom›]
        	            	trace:
        	            	     0.000ms      0.000ms    === operation:process replica _unfinished:‹1› _verbose:‹1› node:‹1› store:‹1› range:‹75/1:/{Table/Max-Max}›
        	            	     0.102ms      0.102ms    event:kv/kvserver/allocator/plan/replicate.go:224 [n1,s1,r75/1:/{Table/Max-Max}] planning range change desc=r75:/{Table/Max-Max} [(n1,s1):1, (n2,s2):2, (n3,s3):3, next=4, gen=5, sticky=9223372036.854775807,2147483647] config=‹range_min_bytes:134217728 range_max_bytes:536870912 gc_policy:<ttl_seconds:14400 ignore_strict_enforcement:true > num_replicas:3 rangefeed_enabled:true ›
        	            	     0.151ms      0.049ms    event:kv/kvserver/allocator/allocatorimpl/allocator.go:1030 [n1,s1,r75/1:/{Table/Max-Max}] replace decommissioning voter - replacement for 1 decommissioning voters priority=5000.00
        	            	     0.171ms      0.020ms    event:kv/kvserver/allocator/plan/replicate.go:235 [n1,s1,r75/1:/{Table/Max-Max}] next replica action: replace decommissioning voter
        	            	     0.237ms      0.066ms    event:kv/kvserver/allocator/allocatorimpl/allocator.go:1481 [n1,s1,r75/1:/{Table/Max-Max}] allocate voter: []" to match "change replicas \(add.*remove.*\): existing descriptor"
        	Test:       	TestReplicateQueueTracingOnError
    panic.go:635: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/bb83881c24a4b108656da9268f863369/logTestReplicateQueueTracingOnError239442307
--- FAIL: TestReplicateQueueTracingOnError (5.10s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #158465 kv/kvserver: TestReplicateQueueTracingOnError failed [A-testing C-bug C-test-failure O-robot P-3 T-kv branch-master]
</p>
</details>
/cc @cockroachdb/kv-triage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestReplicateQueueTracingOnError.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-57889

--- a/pkg/kv/kvserver/BUILD.bazel
+++ b/pkg/kv/kvserver/BUILD.bazel
@@ -391,6 +391,7 @@ go_test(
         "store_test.go",
         "stores_test.go",
         "testutils_test.go",
+        "trace_collection_test.go",
         "ts_maintenance_queue_test.go",
         "txn_recovery_integration_test.go",
         "txn_wait_queue_test.go",

--- a/pkg/kv/kvserver/client_split_test.go
+++ b/pkg/kv/kvserver/client_split_test.go
@@ -12,7 +12,6 @@ import (
 	"math"
 	"math/rand"
 	"reflect"
-	"regexp"
 	"sort"
 	"strconv"
 	"sync/atomic"
@@ -67,13 +66,11 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
-	"github.com/cockroachdb/cockroach/pkg/util/log/logpb"
 	"github.com/cockroachdb/cockroach/pkg/util/protoutil"
 	"github.com/cockroachdb/cockroach/pkg/util/randutil"
 	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
 	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/cockroach/pkg/util/tracing"
-	"github.com/cockroachdb/cockroach/pkg/util/tracing/tracingpb"
 	"github.com/cockroachdb/errors"
 	"github.com/gogo/protobuf/proto"
 	"github.com/stretchr/testify/require"
@@ -1078,123 +1075,49 @@ func TestStoreRangeSplitWithConcurrentWrites(t *testing.T) {
 	})
 }
 
-// TestStoreRangeSplitWithTracing tests that the split queue logs traces for
-// slow splits.
-func TestStoreRangeSplitWithTracing(t *testing.T) {
+// TestSplitTracing verifies that tracing works through the split path by
+// performing a split with a tracing context and verifying child spans are
+// recorded.
+func TestSplitTracing(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
-	l := log.ScopeWithoutShowLogs(t)
-	defer l.Close(t)
-	testutils.SetVModule(t, "split_queue=1")
-
-	splitKey := roachpb.Key("b")
-	var targetRange atomic.Int32
-	manualClock := hlc.NewHybridManualClock()
-	filter := func(ctx context.Context, request *kvpb.BatchRequest) *kvpb.Error {
-		if req, ok := request.GetArg(kvpb.EndTxn); ok {
-			et := req.(*kvpb.EndTxnRequest)
-			if tr := et.InternalCommitTrigger.GetSplitTrigger(); tr != nil {
-				if tr.RightDesc.StartKey.Equal(splitKey) {
-					// Manually increment the replica's clock to simulate a slow split.
-					manualClock.Increment(kvserver.SlowSplitTracingThreshold.Default().Nanoseconds())
-				}
-			}
-		}
-		return nil
-	}
-
-	// Override the load-based split key funciton to force the range to be
-	// processed by the split queue.
-	overrideLBSplitFn := func(rangeID roachpb.RangeID) (roachpb.Key, bool) {
-		if rangeID == roachpb.RangeID(targetRange.Load()) {
-			return splitKey, true
-		}
-		return nil, false
-	}
 
 	ctx := context.Background()
-	s := serverutils.StartServerOnly(t, base.TestServerArgs{
-		Knobs: base.TestingKnobs{
-			Server: &server.TestingKnobs{
-				WallClock: manualClock,
-			},
-			Store: &kvserver.StoreTestingKnobs{
-				DisableMergeQueue:             true,
-				DisableSplitQueue:             true,
-				TestingRequestFilter:          filter,
-				LoadBasedSplittingOverrideKey: overrideLBSplitFn,
-			},
-		},
-	})
-
+	s := serverutils.StartServerOnly(t, base.TestServerArgs{})
 	defer s.Stopper().Stop(ctx)
+
 	store, err := s.GetStores().(*kvserver.Stores).GetStore(s.GetFirstStoreID())
 	require.NoError(t, err)
 
-	// Write some data on both sides of the split key.
+	// Write some data to have something to split.
+	splitKey := roachpb.Key("b")
 	_, pErr := kv.SendWrapped(ctx, store.TestSender(), putArgs([]byte("a"), []byte("foo")))
 	require.NoError(t, pErr.GoError())
 	_, pErr = kv.SendWrapped(ctx, store.TestSender(), putArgs(splitKey, []byte("bar")))
 	require.NoError(t, pErr.GoError())
-	_, pErr = kv.SendWrapped(ctx, store.TestSender(), putArgs([]byte("c"), []byte("foo")))
-	require.NoError(t, pErr.GoError())
 
-	splitKeyAddr, err := keys.Addr(splitKey)
-	require.NoError(t, err)
-	repl := store.LookupReplica(splitKeyAddr)
-	targetRange.Store(int32(repl.RangeID))
-
-	traceCtx, rec := tracing.ContextWithRecordingSpan(ctx, store.GetStoreConfig().Tracer(), "trace-enqueue")
-	processErr, enqueueErr := store.Enqueue(
-		traceCtx, "split", repl, true /* skipShouldQueue */, false, /* async */
+	// Create tracing span for the split.
+	tracer := s.TracerI().(*tracing.Tracer)
+	traceCtx, finishAndGetRecording := tracing.ContextWithRecordingSpan(
+		ctx, tracer, "test-split",
 	)
-	recording := rec()
-	require.NoError(t, enqueueErr)
-	require.NoError(t, processErr)
 
-	// Flush logs and get log messages from split_queue.go
-	log.FlushFiles()
-	entries, err := log.FetchEntriesFromFiles(math.MinInt64, math.MaxInt64, 100,
-		regexp.MustCompile(`split_queue\.go`), log.WithMarkedSensitiveData)
+	// Perform the split with tracing context.
+	err = store.DB().AdminSplit(traceCtx, splitKey, hlc.MaxTimestamp)
 	require.NoError(t, err)
+	recording := finishAndGetRecording()
 
-	opName := "split"
-	traceRegexp, err := regexp.Compile(`trace:.*`)
-	require.NoError(t, err)
-	opRegexp, err := regexp.Compile(fmt.Sprintf(`operation:%s`, opName))
-	require.NoError(t, err)
+	// Sanity checks on the recording.
+	require.NotNil(t, recording)
+	require.GreaterOrEqual(t, recording.Len(), 1)
 
-	// Find the log entry to validate the trace output.
-	foundEntry := false
-	var entry logpb.Entry
-	for _, entry = range entries {
-		if opRegexp.MatchString(entry.Message) {
-			foundEntry = true
-			break
-		}
-	}
-	require.True(t, foundEntry)
+	// Should find our root span.
+	_, found := recording.FindSpan("test-split")
+	require.True(t, found)
 
-	// Validate that the trace is included in the log message.
-	require.Regexp(t, traceRegexp, entry.Message)
-
-	// Validate that the returned tracing span includes the operation, but also
-	// that the stringified trace was not logged to the span or its parent.
-	processRecSpan, foundSpan := recording.FindSpan(opName)
-	require.True(t, foundSpan)
-
-	foundParent := false
-	var parentRecSpan tracingpb.RecordedSpan
-	for _, parentRecSpan = range recording {
-		if parentRecSpan.SpanID == processRecSpan.ParentSpanID {
-			foundParent = true
-			break
-		}
-	}
-	require.True(t, foundParent)
-	spans := tracingpb.Recording{parentRecSpan, processRecSpan}
-	stringifiedSpans := spans.String()
-	require.NotRegexp(t, traceRegexp, stringifiedSpans)
+	// Verify the trace contains expected split events.
+	traceStr := recording.String()
+	require.Regexp(t, `initiating a split`, traceStr)
 }
 
 // TestStoreRangeSplitWithMismatchedDesc ensures that if a RecomputeRequest

--- a/pkg/kv/kvserver/helpers_test.go
+++ b/pkg/kv/kvserver/helpers_test.go
@@ -49,9 +49,15 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/metric"
 	"github.com/cockroachdb/cockroach/pkg/util/quotapool"
 	"github.com/cockroachdb/cockroach/pkg/util/randutil"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing/tracingpb"
 	"github.com/cockroachdb/errors"
 	"github.com/stretchr/testify/require"
 )
+
+// FilterTracingSpans exposes filterTracingSpans for testing.
+func FilterTracingSpans(rec tracingpb.Recording, opNamesToFilter ...string) tracingpb.Recording {
+	return filterTracingSpans(rec, opNamesToFilter...)
+}
 
 func (s *Store) Transport() *RaftTransport {
 	return s.cfg.Transport

--- a/pkg/kv/kvserver/replicate_queue.go
+++ b/pkg/kv/kvserver/replicate_queue.go
@@ -785,6 +785,35 @@ func (decommissionPurgatoryError) PurgatoryErrorMarker() {}
 
 var _ PurgatoryError = decommissionPurgatoryError{}
 
+// withTraceCollection runs fn within a tracing span that is a child of any
+// span in ctx. If recordVerbose is true, verbose recording is enabled and
+// the recording is returned; otherwise the returned recording is nil.
+//
+// The bool return from fn is passed through unchanged. This could be made
+// generic in future if needed for other return types.
+func withTraceCollection(
+	ctx context.Context,
+	tracer *tracing.Tracer,
+	spanName string,
+	recordVerbose bool,
+	fn func(ctx context.Context) (bool, error),
+) (bool, tracingpb.Recording, error) {
+	var opts []tracing.SpanOption
+	if recordVerbose {
+		opts = append(opts, tracing.WithRecording(tracingpb.RecordingVerbose))
+	}
+	ctx, sp := tracing.EnsureChildSpan(ctx, tracer, spanName, opts...)
+	defer sp.Finish()
+
+	result, err := fn(ctx)
+
+	var rec tracingpb.Recording
+	if recordVerbose {
+		rec = sp.GetConfiguredRecording()
+	}
+	return result, rec, err
+}
+
 // filterTracingSpans is a utility for processOneChangeWithTracing in order to
 // remove spans with Operation names in opNamesToFilter, as well as all of
 // their child spans, to exclude overly verbose spans prior to logging.
@@ -820,46 +849,35 @@ func (rq *replicateQueue) processOneChangeWithTracing(
 	priorityAtEnqueue float64,
 ) (requeue bool, _ error) {
 	processStart := timeutil.Now()
-	startTracing := log.ExpensiveLogEnabled(ctx, 1)
-	var opts []tracing.SpanOption
-	if startTracing {
-		// If we enable expensive logging, we also want to record the traces for
-		// the entire operation. We only log the trace below if we both exceed
-		// the timeout and expensive logging is enabled.
-		opts = append(opts, tracing.WithRecording(tracingpb.RecordingVerbose))
-	}
-	ctx, sp := tracing.EnsureChildSpan(ctx, rq.Tracer, "process replica", opts...)
-	defer sp.Finish()
+	recordVerbose := log.ExpensiveLogEnabled(ctx, 1)
 
-	requeue, err := rq.processOneChange(ctx, repl, desc, conf,
-		false /* scatter */, false /* dryRun */, priorityAtEnqueue,
+	requeue, rec, err := withTraceCollection(ctx, rq.Tracer, "process replica", recordVerbose,
+		func(ctx context.Context) (bool, error) {
+			return rq.processOneChange(ctx, repl, desc, conf,
+				false /* scatter */, false /* dryRun */, priorityAtEnqueue,
+			)
+		},
 	)
+
+	// Determine if we should log the trace.
 	processDuration := timeutil.Since(processStart)
 	loggingThreshold := rq.logTracesThresholdFunc(rq.store.cfg.Settings, repl)
-	exceededDuration := loggingThreshold > time.Duration(0) && processDuration > loggingThreshold
+	exceededDuration := loggingThreshold > 0 && processDuration > loggingThreshold
 
-	var traceOutput redact.RedactableString
-	if startTracing {
-		// Utilize a new background context (properly annotated) to avoid writing
-		// traces from a child context into its parent.
-		ctx = repl.AnnotateCtx(rq.AnnotateCtx(context.Background()))
-		var rec tracingpb.Recording
-		traceLoggingNeeded := (err != nil || exceededDuration)
-		if traceLoggingNeeded {
-			// If we have tracing spans from execChangeReplicasTxn, filter it from
-			// the recording so that we can render the traces to the log without it,
-			// as the traces from this span (and its children) are highly verbose.
-			rec = filterTracingSpans(sp.GetConfiguredRecording(),
-				replicaChangeTxnGetDescOpName, replicaChangeTxnUpdateDescOpName,
-			)
-			traceOutput = redact.Sprintf("\ntrace:\n%s", rec)
+	if recordVerbose && (err != nil || exceededDuration) {
+		// Use fresh context to avoid duplicating trace into parent span.
+		logCtx := repl.AnnotateCtx(rq.AnnotateCtx(context.Background()))
+		// Filter out overly verbose spans from execChangeReplicasTxn.
+		rec = filterTracingSpans(rec,
+			replicaChangeTxnGetDescOpName, replicaChangeTxnUpdateDescOpName,
+		)
+		traceOutput := redact.Sprintf("\ntrace:\n%s", rec)
+		if err != nil {
+			log.KvDistribution.Infof(logCtx, "error processing replica: %v%s", err, traceOutput)
+		} else {
+			log.KvDistribution.Infof(logCtx, "processing replica took %s, exceeding threshold of %s%s",
+				processDuration, loggingThreshold, traceOutput)
 		}
-	}
-	if err != nil {
-		log.KvDistribution.Infof(ctx, "error processing replica: %v%s", err, traceOutput)
-	} else if exceededDuration {
-		log.KvDistribution.Infof(ctx, "processing replica took %s, exceeding threshold of %s%s",
-			processDuration, loggingThreshold, traceOutput)
 	}
 
 	return requeue, err

--- a/pkg/kv/kvserver/replicate_queue_test.go
+++ b/pkg/kv/kvserver/replicate_queue_test.go
@@ -50,10 +50,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
-	"github.com/cockroachdb/cockroach/pkg/util/log/logpb"
-	"github.com/cockroachdb/cockroach/pkg/util/timeutil"
 	"github.com/cockroachdb/cockroach/pkg/util/tracing"
-	"github.com/cockroachdb/cockroach/pkg/util/tracing/tracingpb"
 	"github.com/cockroachdb/errors"
 	"github.com/gogo/protobuf/proto"
 	"github.com/stretchr/testify/require"
@@ -877,133 +874,62 @@ func TestReplicateQueueDecommissioningNonVoters(t *testing.T) {
 	})
 }
 
-// TestReplicateQueueTracingOnError tests that an error or slowdown in
-// processing a replica results in traces being logged.
-func TestReplicateQueueTracingOnError(t *testing.T) {
+// TestReplicateQueueTracing is a simple test that verifies tracing works
+// through the replica change path by upreplicating a range with a tracing
+// context and checking that child spans are recorded.
+func TestReplicateQueueTracing(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	s := log.ScopeWithoutShowLogs(t)
-	defer s.Close(t)
-	testutils.SetVModule(t, "replicate_queue=2")
+	defer log.Scope(t).Close(t)
 
-	// NB: This test injects a fake failure during replica rebalancing, and we use
-	// this `rejectSnapshots` variable as a flag to activate or deactivate that
-	// injected failure.
-	var rejectSnapshots int64
 	ctx := context.Background()
-	tc := testcluster.StartTestCluster(
-		t, 4, base.TestClusterArgs{
-			ReplicationMode: base.ReplicationManual,
-			ServerArgs: base.TestServerArgs{Knobs: base.TestingKnobs{Store: &kvserver.StoreTestingKnobs{
-				ReceiveSnapshot: func(_ context.Context, _ *kvserverpb.SnapshotRequest_Header) error {
-					if atomic.LoadInt64(&rejectSnapshots) == 1 {
-						return errors.Newf("boom")
-					}
-					return nil
-				},
-			}}},
-		},
-	)
+	tc := testcluster.StartTestCluster(t, 2, base.TestClusterArgs{
+		ReplicationMode: base.ReplicationManual,
+	})
 	defer tc.Stopper().Stop(ctx)
 
-	// Add a replica to the second and third nodes, and then decommission the
-	// second node. Since there are only 4 nodes in the cluster, the
-	// decommissioning replica must be rebalanced to the fourth node.
-	const decomNodeIdx = 1
-	const decomNodeID = 2
+	// Create scratch range on n1.
 	scratchKey := tc.ScratchRange(t)
-	tc.AddVotersOrFatal(t, scratchKey, tc.Target(decomNodeIdx))
-	tc.AddVotersOrFatal(t, scratchKey, tc.Target(decomNodeIdx+1))
-	adminSrv := tc.Server(decomNodeIdx)
-	adminClient := adminSrv.GetAdminClient(t)
-	_, err := adminClient.Decommission(
-		ctx, &serverpb.DecommissionRequest{
-			NodeIDs:          []roachpb.NodeID{decomNodeID},
-			TargetMembership: livenesspb.MembershipStatus_DECOMMISSIONING,
-		},
+
+	// Create tracing span for the upreplication.
+	tracer := tc.Server(0).TracerI().(*tracing.Tracer)
+	traceCtx, finishAndGetRecording := tracing.ContextWithRecordingSpan(
+		ctx, tracer, "test-upreplication",
 	)
-	require.NoError(t, err)
 
-	// Activate the above testing knob to start rejecting future rebalances and
-	// then attempt to rebalance the decommissioning replica away. We expect a
-	// purgatory error to be returned here.
-	atomic.StoreInt64(&rejectSnapshots, 1)
-	store := tc.GetFirstStoreFromServer(t, 0)
-	repl, err := store.GetReplica(tc.LookupRangeOrFatal(t, scratchKey).RangeID)
-	require.NoError(t, err)
+	// Upreplicate from n1 to n2.
+	tc.AddVotersOrFatalCtx(traceCtx, t, scratchKey, tc.Target(1))
+	recording := finishAndGetRecording()
 
-	testStartTs := timeutil.Now()
-	traceCtx, rec := tracing.ContextWithRecordingSpan(ctx, store.GetStoreConfig().Tracer(), "trace-enqueue")
-	processErr, enqueueErr := tc.GetFirstStoreFromServer(t, 0).Enqueue(
-		traceCtx, "replicate", repl, true /* skipShouldQueue */, false, /* async */
+	// Sanity checks on the recording.
+	require.NotNil(t, recording)
+	require.GreaterOrEqual(t, recording.Len(), 1)
+
+	// Should find our root span.
+	_, found := recording.FindSpan("test-upreplication")
+	require.True(t, found)
+
+	// Should have child spans from the replica change operation.
+	require.Greater(t, len(recording), 1, "expected child spans from upreplication")
+
+	// Verify the trace contains expected replica change events.
+	traceStr := recording.String()
+	require.Regexp(t, `change replicas updated descriptor`, traceStr)
+
+	// The raw recording includes verbose operations that are filtered out by
+	// filterTracingSpans when logging in processOneChangeWithTracing.
+	require.Regexp(t, `operation:change-replica-get-desc`, traceStr)
+	require.Regexp(t, `operation:change-replica-update-desc`, traceStr)
+
+	// Test that filterTracingSpans correctly removes verbose spans.
+	filtered := kvserver.FilterTracingSpans(recording,
+		"change-replica-get-desc", "change-replica-update-desc",
 	)
-	recording := rec()
-	require.NoError(t, enqueueErr)
-	require.Error(t, processErr, "expected processing error")
-
-	// Flush logs and get log messages from replicate_queue.go since just
-	// before calling store.Enqueue(..).
-	log.FlushFiles()
-	entries, err := log.FetchEntriesFromFiles(testStartTs.UnixNano(),
-		math.MaxInt64, 100, regexp.MustCompile(`replicate_queue\.go`), log.WithMarkedSensitiveData)
-	require.NoError(t, err)
-
-	opName := "process replica"
-	errRegexp, err := regexp.Compile(`error processing replica:.*boom`)
-	require.NoError(t, err)
-	traceRegexp, err := regexp.Compile(`trace:.*`)
-	require.NoError(t, err)
-	opRegexp, err := regexp.Compile(fmt.Sprintf(`operation:%s`, opName))
-	require.NoError(t, err)
-
-	// Validate that the error is logged, so that we can use the log entry to
-	// validate the trace output.
-	foundEntry := false
-	var entry logpb.Entry
-	for _, entry = range entries {
-		if errRegexp.MatchString(entry.Message) {
-			foundEntry = true
-			break
-		}
-	}
-	require.True(t, foundEntry)
-
-	// Validate that the trace is included in the log message.
-	require.Regexp(t, traceRegexp, entry.Message)
-	require.Regexp(t, opRegexp, entry.Message)
-
-	// Validate that the logged trace filtered out the verbose execChangeReplicasTxn
-	// child span, as well as the verbose child spans tracing txn operations.
-	require.NotRegexp(t, `operation:change-replica-update-desc`, entry.Message)
-	require.NotRegexp(t, `operation:txn coordinator send`, entry.Message)
-	require.NotRegexp(t, `operation:log-range-event`, entry.Message)
-
-	// Validate that the logged trace includes the changes to the descriptor.
-	require.Regexp(t, `change replicas \(add.*remove.*\): existing descriptor`, entry.Message)
-
-	// Validate that the trace was logged with the correct tags for the replica.
-	require.Regexp(t, fmt.Sprintf("n%d", repl.NodeID()), entry.Tags)
-	require.Regexp(t, fmt.Sprintf("s%d", repl.StoreID()), entry.Tags)
-	require.Regexp(t, fmt.Sprintf("r%d/%d", repl.GetRangeID(), repl.ReplicaID()), entry.Tags)
-	require.Regexp(t, `replicate`, entry.Tags)
-
-	// Validate that the returned tracing span includes the operation, but also
-	// that the stringified trace was not logged to the span or its parent.
-	processRecSpan, foundSpan := recording.FindSpan(opName)
-	require.True(t, foundSpan)
-
-	foundParent := false
-	var parentRecSpan tracingpb.RecordedSpan
-	for _, parentRecSpan = range recording {
-		if parentRecSpan.SpanID == processRecSpan.ParentSpanID {
-			foundParent = true
-			break
-		}
-	}
-	require.True(t, foundParent)
-	spans := tracingpb.Recording{parentRecSpan, processRecSpan}
-	stringifiedSpans := spans.String()
-	require.NotRegexp(t, errRegexp, stringifiedSpans)
-	require.NotRegexp(t, traceRegexp, stringifiedSpans)
+	filteredStr := filtered.String()
+	// The expected message should still be present after filtering.
+	require.Regexp(t, `change replicas updated descriptor`, filteredStr)
+	// The verbose operations should be removed.
+	require.NotRegexp(t, `operation:change-replica-get-desc`, filteredStr)
+	require.NotRegexp(t, `operation:change-replica-update-desc`, filteredStr)
 }
 
 // TestReplicateQueueDecommissionPurgatoryError tests that failure to move a

--- a/pkg/kv/kvserver/split_queue.go
+++ b/pkg/kv/kvserver/split_queue.go
@@ -21,8 +21,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/humanizeutil"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/metric"
-	"github.com/cockroachdb/cockroach/pkg/util/tracing"
-	"github.com/cockroachdb/cockroach/pkg/util/tracing/tracingpb"
 	"github.com/cockroachdb/errors"
 	"github.com/cockroachdb/redact"
 )
@@ -241,35 +239,28 @@ func (sq *splitQueue) processAttemptWithTracing(
 	ctx context.Context, r *Replica, confReader spanconfig.StoreReader,
 ) (processed bool, _ error) {
 	processStart := r.Clock().PhysicalTime()
-	startTracing := log.ExpensiveLogEnabled(ctx, 1)
-	var opts []tracing.SpanOption
-	if startTracing {
-		opts = append(opts, tracing.WithRecording(tracingpb.RecordingVerbose))
-	}
-	ctx, sp := tracing.EnsureChildSpan(ctx, sq.Tracer, "split", opts...)
-	defer sp.Finish()
+	recordVerbose := log.ExpensiveLogEnabled(ctx, 1)
 
-	processed, err := sq.processAttempt(ctx, r, confReader)
+	processed, rec, err := withTraceCollection(ctx, sq.Tracer, "split", recordVerbose,
+		func(ctx context.Context) (bool, error) {
+			return sq.processAttempt(ctx, r, confReader)
+		},
+	)
+
+	// Determine if we should log the trace.
 	processDuration := r.Clock().PhysicalTime().Sub(processStart)
-	exceededDuration := sq.logTracesThreshold > time.Duration(0) && processDuration > sq.logTracesThreshold
-	var traceOutput redact.RedactableString
-	if startTracing {
-		// Utilize a new background context (properly annotated) to avoid writing
-		// traces from a child context into its parent.
-		ctx = r.AnnotateCtx(sq.AnnotateCtx(context.Background()))
+	exceededDuration := sq.logTracesThreshold > 0 && processDuration > sq.logTracesThreshold
 
-		traceLoggingNeeded := (err != nil || exceededDuration)
-		if traceLoggingNeeded {
-			// Add any trace filtering here if the output is too verbose.
-			rec := sp.GetConfiguredRecording()
-			traceOutput = redact.Sprintf("\ntrace:\n%s", rec)
+	if recordVerbose && (err != nil || exceededDuration) {
+		// Use fresh context to avoid duplicating trace into parent span.
+		logCtx := r.AnnotateCtx(sq.AnnotateCtx(context.Background()))
+		traceOutput := redact.Sprintf("\ntrace:\n%s", rec)
+		if err != nil {
+			log.KvDistribution.Infof(logCtx, "error during range split: %v%s", err, traceOutput)
+		} else {
+			log.KvDistribution.Infof(logCtx, "range split took %s, exceeding threshold of %s%s",
+				processDuration, sq.logTracesThreshold, traceOutput)
 		}
-	}
-	if err != nil {
-		log.KvDistribution.Infof(ctx, "error during range split: %v%s", err, traceOutput)
-	} else if exceededDuration {
-		log.KvDistribution.Infof(ctx, "range split took %s, exceeding threshold of %s%s",
-			processDuration, sq.logTracesThreshold, traceOutput)
 	}
 
 	return processed, err

--- a/pkg/kv/kvserver/trace_collection_test.go
+++ b/pkg/kv/kvserver/trace_collection_test.go
@@ -1,0 +1,89 @@
+// Copyright 2024 The Cockroach Authors.
+//
+// Use of this software is governed by the CockroachDB Software License
+// included in the /LICENSE file.
+
+package kvserver
+
+import (
+	"context"
+	"testing"
+
+	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
+	"github.com/cockroachdb/cockroach/pkg/util/log"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing"
+	"github.com/cockroachdb/cockroach/pkg/util/tracing/tracingpb"
+	"github.com/cockroachdb/errors"
+	"github.com/stretchr/testify/require"
+)
+
+func TestWithTraceCollection(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+	tr := tracing.NewTracer()
+
+	t.Run("recordVerbose=true returns recording", func(t *testing.T) {
+		result, rec, err := withTraceCollection(ctx, tr, "test-span", true,
+			func(ctx context.Context) (bool, error) {
+				sp := tracing.SpanFromContext(ctx)
+				require.NotNil(t, sp)
+				require.Equal(t, tracingpb.RecordingVerbose, sp.RecordingType())
+				sp.Record("test message")
+				return true, nil
+			},
+		)
+
+		require.NoError(t, err)
+		require.True(t, result)
+		require.NotNil(t, rec)
+		require.GreaterOrEqual(t, rec.Len(), 1)
+		// Verify the recording contains our span and message.
+		found := false
+		for _, span := range rec {
+			if span.Operation == "test-span" {
+				found = true
+				break
+			}
+		}
+		require.True(t, found, "expected to find test-span in recording")
+	})
+
+	t.Run("recordVerbose=false returns nil recording", func(t *testing.T) {
+		result, rec, err := withTraceCollection(ctx, tr, "test-span", false,
+			func(ctx context.Context) (bool, error) {
+				return false, nil
+			},
+		)
+
+		require.NoError(t, err)
+		require.False(t, result)
+		require.Nil(t, rec)
+	})
+
+	t.Run("error passthrough", func(t *testing.T) {
+		testErr := errors.New("test error")
+		result, rec, err := withTraceCollection(ctx, tr, "test-span", true,
+			func(ctx context.Context) (bool, error) {
+				return false, testErr
+			},
+		)
+
+		require.ErrorIs(t, err, testErr)
+		require.False(t, result)
+		require.NotNil(t, rec, "recording should be returned even on error when recordVerbose=true")
+	})
+
+	t.Run("bool result passthrough", func(t *testing.T) {
+		for _, expected := range []bool{true, false} {
+			result, _, err := withTraceCollection(ctx, tr, "test-span", false,
+				func(ctx context.Context) (bool, error) {
+					return expected, nil
+				},
+			)
+			require.NoError(t, err)
+			require.Equal(t, expected, result)
+		}
+	})
+}

--- a/pkg/testutils/testcluster/testcluster.go
+++ b/pkg/testutils/testcluster/testcluster.go
@@ -886,8 +886,17 @@ func (tc *TestCluster) Targets(serverIdxs ...int) []roachpb.ReplicationTarget {
 func (tc *TestCluster) changeReplicas(
 	changeType roachpb.ReplicaChangeType, startKey roachpb.RKey, targets ...roachpb.ReplicationTarget,
 ) (roachpb.RangeDescriptor, error) {
+	return tc.changeReplicasCtx(context.Background(), changeType, startKey, targets...)
+}
+
+// changeReplicasCtx is like changeReplicas but accepts a context.
+func (tc *TestCluster) changeReplicasCtx(
+	ctx context.Context,
+	changeType roachpb.ReplicaChangeType,
+	startKey roachpb.RKey,
+	targets ...roachpb.ReplicationTarget,
+) (roachpb.RangeDescriptor, error) {
 	tc.t.Helper()
-	ctx := context.TODO()
 
 	var returnErr error
 	var desc *roachpb.RangeDescriptor
@@ -932,10 +941,20 @@ func (tc *TestCluster) changeReplicas(
 func (tc *TestCluster) addReplica(
 	startKey roachpb.Key, typ roachpb.ReplicaChangeType, targets ...roachpb.ReplicationTarget,
 ) (roachpb.RangeDescriptor, error) {
+	return tc.addReplicaCtx(context.Background(), startKey, typ, targets...)
+}
+
+// addReplicaCtx is like addReplica but accepts a context.
+func (tc *TestCluster) addReplicaCtx(
+	ctx context.Context,
+	startKey roachpb.Key,
+	typ roachpb.ReplicaChangeType,
+	targets ...roachpb.ReplicationTarget,
+) (roachpb.RangeDescriptor, error) {
 	rKey := keys.MustAddr(startKey)
 
-	rangeDesc, err := tc.changeReplicas(
-		typ, rKey, targets...,
+	rangeDesc, err := tc.changeReplicasCtx(
+		ctx, typ, rKey, targets...,
 	)
 	if err != nil {
 		return roachpb.RangeDescriptor{}, err
@@ -952,7 +971,14 @@ func (tc *TestCluster) addReplica(
 func (tc *TestCluster) AddVoters(
 	startKey roachpb.Key, targets ...roachpb.ReplicationTarget,
 ) (roachpb.RangeDescriptor, error) {
-	return tc.addReplica(startKey, roachpb.ADD_VOTER, targets...)
+	return tc.AddVotersCtx(context.Background(), startKey, targets...)
+}
+
+// AddVotersCtx is like AddVoters but accepts a context.
+func (tc *TestCluster) AddVotersCtx(
+	ctx context.Context, startKey roachpb.Key, targets ...roachpb.ReplicationTarget,
+) (roachpb.RangeDescriptor, error) {
+	return tc.addReplicaCtx(ctx, startKey, roachpb.ADD_VOTER, targets...)
 }
 
 // AddNonVoters is part of TestClusterInterface.
@@ -1075,8 +1101,18 @@ func (tc *TestCluster) waitForNewReplicas(
 func (tc *TestCluster) AddVotersOrFatal(
 	t serverutils.TestFataler, startKey roachpb.Key, targets ...roachpb.ReplicationTarget,
 ) roachpb.RangeDescriptor {
+	return tc.AddVotersOrFatalCtx(context.Background(), t, startKey, targets...)
+}
+
+// AddVotersOrFatalCtx is like AddVotersOrFatal but accepts a context.
+func (tc *TestCluster) AddVotersOrFatalCtx(
+	ctx context.Context,
+	t serverutils.TestFataler,
+	startKey roachpb.Key,
+	targets ...roachpb.ReplicationTarget,
+) roachpb.RangeDescriptor {
 	t.Helper()
-	desc, err := tc.AddVoters(startKey, targets...)
+	desc, err := tc.AddVotersCtx(ctx, startKey, targets...)
 	if err != nil {
 		t.Fatalf(`could not add %v replicas to range containing %s: %+v`,
 			targets, startKey, err)

